awssql/plugins/aurora_initial_connection_strategy_plugin.go

```diff
@@ -0,0 +1,421 @@
+					time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+					continue
+				}
+				return nil, err
+			}
+			readerCandidate, err = plugin.pluginService.IdentifyConnection(readerCandidateConn)
+			if err != nil {
+				if plugin.handleGetVerifiedRoleConnectionError(readerCandidateConn, readerCandidate, err) {
+					time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+					continue
+				}
+				return nil, err
+			}
+
+			if readerCandidate == nil {
+				if readerCandidateConn != nil {
+					_ = readerCandidateConn.Close()
+				}
+				time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+				continue
+			}
+
+			if readerCandidate.Role != host_info_util.READER {
+				if plugin.hasNoReaders() {
+					// It seems that cluster has no readers. Simulate Aurora reader cluster endpoint logic
+					// and return the current (writer) connection.
+					if isInitialConnection {
+						plugin.hostListProviderService.SetInitialConnectionHostInfo(readerCandidate)
+					}
+					return readerCandidateConn, nil
+				}
+				if readerCandidateConn != nil {
+					_ = readerCandidateConn.Close()
+				}
+				time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+				continue
+			}
+
+			if isInitialConnection {
+				plugin.hostListProviderService.SetInitialConnectionHostInfo(readerCandidate)
+			}
+			return readerCandidateConn, nil
+		}
+
+		readerCandidateConn, err = plugin.pluginService.Connect(readerCandidate, props, plugin)
+		if err != nil {
+			if plugin.handleGetVerifiedRoleConnectionError(readerCandidateConn, readerCandidate, err) {
+				time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+				continue
+			}
+			return nil, err
+		}
+
+		if plugin.pluginService.GetHostRole(readerCandidateConn) != host_info_util.READER {
+			// If the new connection resolves to a writer instance, this means the topology is outdated.
+			// Force refresh to update the topology.
+			err = plugin.pluginService.ForceRefreshHostList(readerCandidateConn)
+			if err != nil {
+				if plugin.handleGetVerifiedRoleConnectionError(readerCandidateConn, readerCandidate, err) {
+					time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+					continue
+				}
+				return nil, err
+			}
+
+			if plugin.hasNoReaders() {
+				// It seems that cluster has no readers. Simulate Aurora reader cluster endpoint logic
+				// and return the current (writer) connection.
+				if isInitialConnection {
+					plugin.hostListProviderService.SetInitialConnectionHostInfo(readerCandidate)
+				}
+				return readerCandidateConn, nil
+			}
+			_ = readerCandidateConn.Close()
+			time.Sleep(time.Millisecond * time.Duration(retryDelayMs))
+			continue
+		}
+		if isInitialConnection {
+			plugin.hostListProviderService.SetInitialConnectionHostInfo(readerCandidate)
+		}
+		return readerCandidateConn, nil
+	}
+	return nil, nil
+}
+
+func (plugin *AuroraInitialConnectionStrategyPlugin) handleGetVerifiedRoleConnectionError(
```

Comment:
I don't think this is a great name, reading this at first glance, not sure what the boolean value is supposed to represent. Maybe `handleErrorAndShouldRetry` would be a better name?